src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/persistence/SQLitePersistor.java

```diff
@@ -179,6 +221,426 @@ synchronized int insertRow(String tableName, List<Object> row) throws SQLExcepti
+    }
+    Class mTypeArgClass = (Class) mTypeArguments[0];
+    return mTypeArgClass;
+  }
+
+  private void checkPublic(Method method) {
+    if (!Modifier.isPublic(method.getModifiers())) {
+      throw new IllegalStateException("Found '" + method.getName() + "'. But it is not public.");
+    }
+  }
+
+  private String capitalize(String name) {
+    if (name.length() == 1) {
+      return name.toUpperCase();
+    }
+    return name.substring(0, 1).toUpperCase() + name.substring(1);
+  }
+
+  private void checkValidType(java.lang.reflect.Field field, Class<?> clz) {
+    Type type = field.getGenericType();
+    Annotation[] annotation = field.getAnnotations();
+    StringBuilder err = new StringBuilder();
+    err.append("[class: ").append(clz.getSimpleName()).append(":field:").append(field.getName()).append("]::");
+
+    if (annotation.length != 1) {
+      err
+          .append("A field can have either ")
+          .append(ValueColumn.class.getSimpleName())
+          .append(" or ")
+          .append(RefColumn.class.getSimpleName())
+          .append(" not both");
+      throw new IllegalStateException(err.toString());
+    }
+
+    if (annotation[0].annotationType() == RefColumn.class) {
+      if (type instanceof ParameterizedType) {
+        ParameterizedType parameterizedType = (ParameterizedType) type;
+        Class<?> actualType = (Class<?>) parameterizedType.getActualTypeArguments()[0];
+        if (actualType.isPrimitive() || actualType == String.class) {
+          err.append("RefColumn can only be used for a user defined class or a Collection of user-defined class NOT "
+              + "a primitive or String type");
+          throw new IllegalStateException(err.toString());
+        }
+      } else {
+        Class<?> justType = field.getType();
+        if (justType.isPrimitive() || justType == String.class) {
+          err.append("RefColumn can only be used for a user defined class or a Collection of user-defined class NOT "
+              + "a primitive or String type");
+          throw new IllegalStateException(err.toString());
+        }
+      }
+    } else {
+      if (type instanceof ParameterizedType) {
+        // TODO: This will be added if use-cases require this.
+        err.append("persisting Primitives or Strings as Parameterized Types is not supported.");
+        throw new IllegalStateException(err.toString());
+      }
+    }
+  }
+
+  /**
+   * Go over all the fields of the class and then filter out all that are annotated as @AColumn or @ATable. For those fields,
+   * try to figure out the getter and setters.
+   * @param clz The class whose field registry is to be created.
+   * @param <T> The Generic type of the class.
+   * @throws IllegalStateException When getters and setters are not found for the field that is required to be persisted or they exist but
+   *     are not public.
+   */
+  private <T> void createFieldRegistry(Class<T> clz) throws IllegalStateException, NoSuchMethodException {
```

Comment:
a high level question I have for this function is, how are we going to handle Enum type ? We have resource enum in ResourceSummary. Can we override this getter / setter to serialize / de-serialize Enum based on enum value ?

Reply:
This is an excellent point. Enums are not covered. We will add support for them before we migrate Existing RCAs to be persisted using this new API. 
For now this should unblock the persisting the deciders into DB use-case